staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery.go

```diff
@@ -406,14 +406,38 @@ func (dm *discoveryManager) syncAPIService(apiServiceName string) error {
 	return nil
 }
 
+func (dm *discoveryManager) markInitialStale() {
```

Comment:
Should `AddAPIService` be calling `AddGroupVersion` to add the service as stale if it is unknown?

https://github.com/kubernetes/kubernetes/blob/57b27fd3cd11cb5f2515c7ac5f67f612998fb368/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery.go#L464-L480

Reply:
Not opposed to it. Does it make sense that newly added APIServices (outside of initialization health check period) have a chance to appear as stale even though they're always healthy?

Reply:
Good point, it could be worked around if `discoveryManager` had a `HasSynced` function or some state (to take the place of `discoverySyncedCh`). That way if we had not synced we wouldnt set it as stale. This approach has the benefit of following our other controller patterns (HasSynced() + Run(ctx/stopCh)), but I appreciate to do so might have a lot of ripple and not be worth it right now.

I think both would be equally valid approaches. I'm fine with either one. I'd leave it to you to ruminate/experiment on that and ok with whatever you choose here.

Reply:
I feel like the HasSynced interface is used for waiting on informers which we're not exactly using with this controller so it might make sense to go with a slightly modified interface. My one gripe with having some kind of variable is the need for a lock to read a variable that would only be modified for a small period of time before always being set to true. In addition, AddAPIService is called on a set interval (sync) from the api registration controller, and I don't like that we have to do the synced check (or checking for existence) every time.

I still prefer the current approach where we do something during initialization and not worry about it afterwards. wdyt?

Reply:
sgtm.